operator/v1/types_insights.go

```diff
@@ -0,0 +1,134 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+//
+// InsightsOperator holds cluster-wide information about the Insights Operator.
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type InsightsOperator struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	// spec is the specification of the desired behavior of the Insights.
+	// +kubebuilder:validation:Required
+	Spec InsightsOperatorSpec `json:"spec"`
+
+	// status is the most recently observed status of the Insights operator.
+	// +optional
+	Status InsightsOperatorStatus `json:"status"`
+}
+
+type InsightsOperatorSpec struct {
+	OperatorSpec `json:",inline"`
+}
+
+type InsightsOperatorStatus struct {
+	OperatorStatus `json:",inline"`
+	// gatherStatus provides basic information about the last Insights data gathering.
+	// When omitted, this means no data gathering has taken place yet.
+	// +optional
+	GatherStatus GatherStatus `json:"gatherStatus,omitempty"`
+	// insightsReport provides general Insights analysis results.
+	// +optional
+	InsightsReport InsightsReport `json:"insightsReport,omitempty"`
+}
+
+// gatherStatus provides information about the last known gather event.
+type GatherStatus struct {
+	// lastGatherTime is the last time when Insights data gathering finished.
+	// An empty value means that no data has been gathered yet.
+	// +optional
+	LastGatherTime metav1.Time `json:"lastGatherTime,omitempty"`
+	// lastGatherDuration is the total time taken to process
+	// all gatherers during the last gather event.
+	// +optional
+	LastGatherDuration metav1.Duration `json:"lastGatherDuration,omitempty"`
```

Comment:
We should add some validation here
```suggestion
	// +kubebuilder:validation:Pattern="^0|([0-9]+(\\.[0-9]+)?(ns|us|µs|ms|s|m|h))+$"
	LastGatherDuration metav1.Duration `json:"lastGatherDuration,omitempty"`
```
Note this includes `0` at the front to allow a zero value else the validation would fail when it's omitted.

Reply:
OK thanks.

Reply:
Hmm I am getting:
```
/go/src/github.com/openshift/api/operator/v1/types_insights.go:54:2: must apply pattern to a `string` or `IntOrString`
```
Would it make sense to change it to string?

Reply:
Try adding `// +kubebuilder:validation:Type:=string` and see if that helps, thinking about it, you may not need the `0` 🤔 

Once we work out exactly how these work I'm going to document durations in the API conventions

Reply:
Thanks. It passed the generation.